### ISSUE TITLE
ADL:Issues import stuck in pending

### ISSUE THREAD

**Issue by jillpe** [scientist-softserv/adventist-dl#240]
# Summary
I'm trying to test the adl:issue set import, and all works are stuck pending. https://adl.s2.adventistdigitallibrary.org/importers/35?locale=en

# Accepted Criteria
- [ ] When importing a set, all of the works get uploaded

# Chores

- [ ] Add `Bulkrax::OaiAdventistQdcEntry#collections_created` method that returns `true`
- [ ] Add `Bulkrax::OaiAdventistQdcEntry#find_collection_ids` method that modifies the guard line (e.g. `return self.collection_ids if collections_created?`).

```ruby
def find_collection_ids
  return self.collection_ids if @found_collection_ids

  if sets.blank? || parser.collection_name != 'all'
    collection = find_collection(importerexporter.unique_collection_identifier(parser.collection_name))
    self.collection_ids << collection.id if collection.present? && !self.collection_ids.include?(collection.id)
  else # All - collections should exist for all sets
    sets.each do |set|
      c = find_collection(importerexporter.unique_collection_identifier(set.content))
      self.collection_ids << c.id if c.present? && !self.collection_ids.include?(c.id)
    end
  end
  @found_collection_ids = true

  return self.collection_ids
end
```

Note: There is exceptions being reported to Sentry that are not moving the "pending" status to "Error"; consider resolving that. This would require a Bulkrax update.  However this is out of scope for the ticket.

## NOTES

Bulkrax Issue: https://github.com/samvera-labs/bulkrax/issues/459

## TESTING INSTRUCTIONS

Create and run an importer that looks like the following, in staging. 
- [ ] the entries' statuses should no longer be stuck in a pending state

![image](https://user-images.githubusercontent.com/10081604/216424864-fb251d05-fdbd-429a-9f02-554612342781.png)

**Comment by jillpe** [comment]
refer to the notes on #241

**Comment by jeremyf** [comment]
Baseline assumption of the default OAI Entry:

All collections must exist before you make works.  We decided that when I import a set, only put the works into that set’s representative collection.

In the case of Adventist, there is a collection we don’t need (e.g. “adl:issue”) but we do need the aark based on to already exist.


## Code Sleuthing

The [Bulkrax::ImportBehavior#build\_for\_importer](https://github.com/samvera-labs/bulkrax/blob/5f6b6afa25e494f19a82ff8d7ae51661a9b36c20/app/models/concerns/bulkrax/import_behavior.rb#L8-L28) method raises a `CollectionsCreatedError` when we didn’t create collections.  Each entry is responsible for implementing that method.

<details>
<summary>Bulkrax::ImportBehavior#build\_for\_importer</summary>

```ruby
def build_for_importer
  begin
    build_metadata
    unless self.importerexporter.validate_only
      raise CollectionsCreatedError unless collections_created?
      @item = factory.run!
      add_user_to_permission_templates! if self.class.to_s.include?("Collection")
      parent_jobs if self.parsed_metadata[related_parents_parsed_mapping]&.join.present?
      child_jobs if self.parsed_metadata[related_children_parsed_mapping]&.join.present?
    end
  rescue RSolr::Error::Http, CollectionsCreatedError => e
    raise e
  rescue StandardError => e
    status_info(e)
  else
    status_info
  ensure
    self.save!
  end
  return @item
end
```

</details>

Below is the [Bulkrax::OaiEntry#collections\_created?](https://github.com/samvera-labs/bulkrax/blob/5f6b6afa25e494f19a82ff8d7ae51661a9b36c20/app/models/bulkrax/oai_entry.rb#L51-L57) method definition.  It is what the `Bulkrax::OaiAdventistQdcEntry` uses (which is used by Adventist’s OAI parser).  In the below method we are not concerned with the magic ‘all’ but instead with the case where we say each entry must have one and only one collection.

<details>
<summary>Bulkrax::OaiEntry#collections_created?</summary>

```ruby
def collections_created?
  if parser.collection_name == 'all'
    sets.blank? || (sets.present? && sets.size == self.collection_ids.size)
  else
    self.collection_ids.size == 1
  end
end
```

</details>

What’s the logic for setting collections?  It originates in [Bulkrax::ImportBehavior#add\_collections](https://github.com/samvera-labs/bulkrax/blob/5f6b6afa25e494f19a82ff8d7ae51661a9b36c20/app/models/concerns/bulkrax/import_behavior.rb#L99-L106) which calls `#find_collection_ids`.

<details>
<summary>Bulkrax::ImportBehavior#add_collections</summary>

```ruby
def add_collections
  return if find_collection_ids.blank?
  
  self.parsed_metadata['member_of_collections_attributes'] = {}
  find_collection_ids.each_with_index do |c, i|
    self.parsed_metadata['member_of_collections_attributes'][i.to_s] = { id: c }
  end
end
```

</details>

For [Bulkrax::OaiEntry#find\_collection\_ids](https://github.com/samvera-labs/bulkrax/blob/5f6b6afa25e494f19a82ff8d7ae51661a9b36c20/app/models/bulkrax/oai_entry.rb#L93-L105) we have the following definition.  Again note the ‘all’ string.  Note, we short-circuit the test if we’ve created collections.  This is a bit of a surprise; I would’ve expected a different form of memoization.  It seems like we could enter into an odd state.

<details>
<summary>Bulkrax::OaiEntry#find_collection_ids</summary>

```ruby
def find_collection_ids
  return self.collection_ids if collections_created?
  if sets.blank? || parser.collection_name != 'all'
    collection = find_collection(importerexporter.unique_collection_identifier(parser.collection_name))
    self.collection_ids << collection.id if collection.present? && !self.collection_ids.include?(collection.id)
  else # All - collections should exist for all sets
    sets.each do |set|
      c = find_collection(importerexporter.unique_collection_identifier(set.content))
      self.collection_ids << c.id if c.present? && !self.collection_ids.include?(c.id)
    end
  end
  return self.collection_ids
end
```

</details>

Given that some of Adventists records have two sets (e.g. one that’s an `aark_id` and one that is the name of the OAI set).  If both of those are found they would both be added to the `collection_ids` array; meaning there are 2 entries which would raise an exception.

Digging further, Adventist’s [Bulkrax::HasLocalProcessing#add\_local](https://github.com/scientist-softserv/adventist-dl/blob/14037e60088c97703316a7ed04af1efccc2f1bc7/app/models/concerns/bulkrax/has_local_processing.rb) has logic that can and will create collections.  And some of what’s created could be things other  than the above `collection_ids`; namely those produced through the `part_of` property.

<details>
<summary>Bulkrax::HasLocalProcessing#add_local</summary>

```ruby
module Bulkrax
  module HasLocalProcessing
    # This method is called during build_metadata
    # add any special processing here, for example to reset a metadata property
    # to add a custom property from outside of the import data
    def add_local
      if self.parsed_metadata['model'] == 'Collection' || factory_class == Collection
        add_collection_nesting
      else
        add_part_of_collections if self.parsed_metadata['part_of'].present?
        add_set_collections
      end
    end
    
    def add_set_collections
      # The #header method is for OAI records
      return unless record.respond_to?(:header)
      sets = record.header.set_spec.map(&:content)
      aark_ids = sets.select { |s| s.match(/^\d+$/) }
      aark_ids.each do |aark|
        collection = Collection.where(aark_id_tesim: aark).first
        add_collection_to_work(collection) if collection
      end
    end
    
    def add_part_of_collections
      self.parsed_metadata['part_of'].each do |collection_title|
        collection = find_or_create_collection(collection_title)
        add_collection_to_work(collection)
      end
    end
    
    def add_collection_to_work(collection)
      self.parsed_metadata['member_of_collections_attributes'] ||= {}
      top_key = self.parsed_metadata['member_of_collections_attributes'].keys.map { |k| k.to_i }.sort.last || -1
      self.parsed_metadata['member_of_collections_attributes'][(top_key + 1).to_s] = { id: collection.id }
    end
    
    def add_collection_nesting
      self.parsed_metadata.delete('admin_set_id')
      nested_collections = self.parsed_metadata.delete('member_of_collections_attributes')
      self.parsed_metadata['collection'] = nested_collections.map { |k, v| v[:id] } if nested_collections.is_a?(Hash)
      
      if self.parsed_metadata['part_of'].present?
        self.parsed_metadata['part_of'].each do |collection_title|
          collection = find_or_create_collection(collection_title)
          self.parsed_metadata['collection'] ||= []
          self.parsed_metadata['collection'] << collection.id
        end
      end
    end
    
    def find_or_create_collection(collection_title)
      Collection.where(title_sim: collection_title).first || Collection.create(title: [collection_title], collection_type: Hyrax::CollectionType.find_by(title: 'User Collection'))
    end
  end
end
```

</details>

Looking at [Bulkrax::Entry 187125](https://adl.s2.adventistdigitallibrary.org/importers/35/entries/187125?locale=en), which [raised a Bulkrax::CollectionsCreatedError](https://sentry.io/organizations/scientist-inc/issues/3827889423/?project=6745020&query=is%3Aunresolved&referrer=issue-stream), the persisted `collection_ids` value is an empty array (e.g. `[]`).

```shell
irb(main):007:0> Bulkrax::Entry.find(187125).collection_ids
D, [2023-02-01T16:36:18.120491 #71] DEBUG -- :   Bulkrax::Entry Load (1.1ms)  SELECT  "bulkrax_entries".* FROM "bulkrax_entries" WHERE "bulkrax_entries"."id" = $1 LIMIT $2  [["id", 187125], ["LIMIT", 1]]
=> []
irb(main):008:0>
```

Yet the persisted `parsed_metadata["member_of_collections_attributes"]` is not empty.

```shell
irb(main):009:0> Bulkrax::Entry.find(187125).parsed_metadata["member_of_collections_attributes"]
D, [2023-02-01T16:39:48.973848 #71] DEBUG -- :   Bulkrax::Entry Load (1.0ms)  SELECT  "bulkrax_entries".* FROM "bulkrax_entries" WHERE "bulkrax_entries"."id" = $1 LIMIT $2  [["id", 187125], ["LIMIT", 1]]
=> {"0"=>{"id"=>"ffd7d24c-2d65-4390-8394-6c33e01e2cbe"}}
irb(main):010:0>
```

This is because collections added via `part_of` don’t count towards the `collection_ids` that we’re checking to say `collection_created?`. *Note: In the case of the above entry, it was run when we were parsing title into both the title and part property.  That was addressed in [PR 226](https://github.com/scientist-softserv/adventist-dl/pull/226), which remediates a misreading of [Issue 101](https://github.com/scientist-softserv/adventist-dl/issues/101).*



## Proposed Solution and Request for Documentation

Given the above code and metadata, I recommend that we adjust the Adventist specific logic for `collections_created?` to always return `true`.   I recommend this with trepedation.  The Adventist logic has eschewed the assumption of requiring collections be created before creating works that are part of the collections (via the `find_or_create_collection` logic).  That assumption exists to reduce timing issues (e.g. creating multiple collections with the same name).

I am looking for discussion around this recommendation.

**Comment by bkiahstroud** [comment]
>In the case of Adventist, there is a collection we don’t need (e.g. “adl:issue”) but we do need the aark based on to already exist.

@jeremyf what do you mean when you say "there is a collection we don't need"? As in we don't add the works to that collection?

**Comment by jeremyf** [comment]
> > In the case of Adventist, there is a collection we don’t need (e.g. “adl:issue”) but we do need the aark based on to already exist.
> 
> @jeremyf what do you mean when you say "there is a collection we don't need"? As in we don't add the works to that collection?

@bkiahstroud the "all:issue" is a quasi-collection.  It happens to be the OAI Set's name.  It is not something meaningful to Adventist.  They are prepared to create those collections to move past any blockers, but their presence is a foible of implementation detail for their feed.

**Comment by orangewolf** [comment]
I've read through this whole thing and agree with you that overriding `collections_created?` is a reasonable approach here.

**Comment by ShanaLMoore** [comment]
![image](https://user-images.githubusercontent.com/10081604/216179742-13428125-394b-4bdf-a9eb-a5b3f81866c0.png)

**Comment by ShanaLMoore** [comment]
# QA RESULTS

After deploying, we re ran the offending importer.

Link to importer: https://adl.s2.adventistdigitallibrary.org/importers/35?locale=en

**Comment by KatharineV** [comment]
I created an issue importer on staging, and most of the works have failed with the error message Faraday::ConnectionFailed. https://adl.s2.adventistdigitallibrary.org/importers/43?locale=en

**Comment by DiemBTran** [comment]
> I created an issue importer on staging, and most of the works have failed with the error message Faraday::ConnectionFailed. https://adl.s2.adventistdigitallibrary.org/importers/43?locale=en

Thanks for your QA notes, Katharine! I have added the `Needs Rework` label and pushed this ticket back to `Ready for Dev` for the devs to look at.

**Comment by ShanaLMoore** [comment]
faraday issues may need to be handed off to team teal 6

**Comment by ShanaLMoore** [comment]
With Rob's help, we added nodes to the adl rancher cluster and increased fcrepo resources. 

I re ran two previously offending importers and it appears to be resolved: 

1. [Faraday Timeout](https://adl.s2.adventistdigitallibrary.org/importers/43?locale=en)
2. [LDP::NotFound](https://adl.s2.adventistdigitallibrary.org/importers/35?locale=en#work-entries)

### Screenshots
FARADAY



![Image](https://user-images.githubusercontent.com/10081604/219141642-3979c4fd-afaa-45b2-ad00-9552a4e8a8f1.png)



LDP::NOTFOUND


![Image](https://user-images.githubusercontent.com/10081604/219141583-8ea11a28-b137-4ae4-91ea-228a5ae4b519.png)

**Comment by KatharineV** [comment]
@ShanaLMoore Thanks for the heads-up to rerun a test. I created a [new 100 issue importer](https://adl.s2.adventistdigitallibrary.org/importers/51?locale=en) on staging and ran it today. It's been going for a little under an hour now. Currently, 78 entries are processed and 22 are stuck (?) at pending. I don't see any error messages. I just haven't seen the processed/pending numbers change in a while. I'll keep refreshing and will update here if it changes.

**Comment by KatharineV** [comment]
Hi team. The [importer ](https://adl.s2.adventistdigitallibrary.org/importers/51?locale=en)I created for testing on 2/16 never did move beyond 78 entries processed and 22 stuck in pending. Today I created another 100 issue [importer](https://adl.s2.adventistdigitallibrary.org/importers/60?locale=en), just to make sure I could replicate the results. It's been about 15 minutes, and I'm seeing 65 entries processed and 39 enqueued/pending.
